File: app/src/main/java/com/example/policyfolio/UI/BottomSheets/InfoBottomSheet.java
package com.example.policyfolio.UI.BottomSheets;


import android.annotation.SuppressLint;
import android.app.Dialog;
import android.os.Bundle;
import android.telephony.PhoneNumberFormattingTextWatcher;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.policyfolio.UI.Adapters.ListAdapters.BasicDropdownTextAdapter;
import com.example.policyfolio.Util.Constants;
import com.example.policyfolio.R;
import com.example.policyfolio.ViewModels.WithUser.HomeViewModel;
import com.hbb20.CountryCodePicker;

import java.util.Calendar;

/**
 * A simple {@link Fragment} subclass.
 */
public class InfoBottomSheet extends Fragment implements BasicDropdownTextAdapter.ParentCallback {

    private InfoSheetCallback callback;
    private HomeViewModel viewModel;
    private View rootView;

    private EditText name;
    private EditText email;
    private LinearLayout phone;
    private EditText phoneText;
    private CountryCodePicker ccp;
    private TextView birthday;
    private TextView genderText;
    private RecyclerView genderChoice;
    private EditText city;
    private Button save;

    private TextView nameError;
    private TextView emailError;
    private TextView phoneError;
    private TextView birthdayError;
    private TextView cityError;

    private Long birthdayEpoch;
    private BasicDropdownTextAdapter genderAdapter;
    private String[] genderArray;
    private int genderSelection;

    public InfoBottomSheet() {
    }

    @SuppressLint("ValidFragment")
    public InfoBottomSheet(InfoSheetCallback callback, HomeViewModel viewModel){
        this.callback = callback;
        this.viewModel = viewModel;
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.bottom_sheet_info, container, false);

        name = rootView.findViewById(R.id.name);
        email = rootView.findViewById(R.id.email);
        phoneText = rootView.findViewById(R.id.phone_text);
        phone = rootView.findViewById(R.id.phone);
        ccp = rootView.findViewById(R.id.ccp);
        birthday = rootView.findViewById(R.id.birthday);
        genderText = rootView.findViewById(R.id.gender_text);
        genderChoice = rootView.findViewById(R.id.gender_choice);
        city = rootView.findViewById(R.id.city);
        save = rootView.findViewById(R.id.save);

        nameError = rootView.findViewById(R.id.name_empty);
        emailError = rootView.findViewById(R.id.email_empty);
        phoneError = rootView.findViewById(R.id.phone_empty);
        birthdayError = rootView.findViewById(R.id.birthday_empty);
        cityError = rootView.findViewById(R.id.city_empty);

        ccp.registerCarrierNumberEditText(phoneText);
        ccp.setNumberAutoFormattingEnabled(true);

        setGenderAdapter();

        birthday.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                FetchDate();
            }
        });

        save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                save();
            }
        });

        setAvailableInfo();

        return rootView;
    }

    public void setAvailableInfo() {
        if(viewModel.getEmail()!=null)
            email.setText(viewModel.getEmail());
        if(viewModel.getName()!=null)
            name.setText(viewModel.getName());
        if(viewModel.getPhone()!= null)
            phoneText.setText(viewModel.getPhone());
        if (viewModel.getBirthday()!=null) {
            birthday.setText(Constants.Time.DATE_FORMAT.format(viewModel.getBirthday()*1000));
            birthdayEpoch = viewModel.getBirthday();
            birthday.setTextColor(getResources().getColor(R.color.white));
        }
        setValue(viewModel.getGender(), Constants.ListTypes.GENDER);
        if(viewModel.getCity()!=null)
            city.setText(viewModel.getCity());
    }

    private void save() {
        String name = this.name.getText().toString();
        if(name.equals("")){
            nameError.setVisibility(View.VISIBLE);
        }
        else {
            nameError.setVisibility(View.GONE);
            viewModel.setName(name);
        }

        String email = this.email.getText().toString();
        if (email.equals("")) {
            emailError.setVisibility(View.VISIBLE);
            emailError.setText("The field cannot be left empty");
        }
        else if(!Patterns.EMAIL_ADDRESS.matcher(email).matches()){
            emailError.setVisibility(View.VISIBLE);
            emailError.setText("Invalid Email Id");
        }
        else {
            emailError.setVisibility(View.GONE);
            viewModel.setEmail(email);
        }

        String phone = ccp.getFormattedFullNumber();
        if (phone.equals("")) {
            phoneError.setVisibility(View.VISIBLE);
            phoneError.setText("Empty Field Not Allowed");
        }
        else if(!ccp.isValidFullNumber()){
            phoneError.setVisibility(View.VISIBLE);
            phoneError.setText("Invalid PhoneNumber");
        }
        else {
            phoneError.setVisibility(View.GONE);
            viewModel.setPhone(phone);
        }

        if(birthdayEpoch!=null){
            birthdayError.setVisibility(View.GONE);
            viewModel.setBirthday(birthdayEpoch);
        }
        else {
            birthdayError.setVisibility(View.VISIBLE);
        }

        viewModel.setGender(genderSelection);

        String city = this.city.getText().toString();
        if(city.equals("")){
            cityError.setVisibility(View.VISIBLE);
        }
        else {
            cityError.setVisibility(View.GONE);
            viewModel.setCity(city);
        }


        if(nameError.getVisibility() == View.GONE && birthdayError.getVisibility() == View.GONE && cityError.getVisibility() == View.GONE && phoneError.getVisibility() == View.GONE && emailError.getVisibility() == View.GONE)
            callback.updateInfo();
    }

    private void FetchDate() {
        final Dialog dialog = new Dialog(getContext());
        dialog.setContentView(R.layout.date_picker);
        dialog.setTitle("");
        DatePicker datePicker = dialog.findViewById(R.id.date_picker);
        final Calendar calendar=Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        datePicker.init(calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH), calendar.get(Calendar.DAY_OF_MONTH), new DatePicker.OnDateChangedListener() {
            @Override
            public void onDateChanged(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                calendar.set(year,monthOfYear,dayOfMonth);
                birthdayEpoch = (calendar.getTimeInMillis())/1000;
                birthday.setText(Constants.Time.DATE_FORMAT.format(birthdayEpoch*1000));
                birthday.setTextColor(getResources().getColor(R.color.white));
            }
        });
        dialog.show();
    }

    private void setGenderAdapter() {
        genderSelection = 0;
        genderArray = getResources().getStringArray(R.array.gender_array);
        genderAdapter = new BasicDropdownTextAdapter(getContext(),genderArray,this, Constants.ListTypes.GENDER, getResources().getColor(R.color.colorPrimaryDarkest));
        genderChoice.setAdapter(genderAdapter);
        genderChoice.setLayoutManager(new LinearLayoutManager(getContext(),RecyclerView.VERTICAL,false));

        genderText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                genderText.setVisibility(View.GONE);
                genderChoice.setVisibility(View.VISIBLE);
            }
        });
    }

    @Override
    public void setValue(int position, int type) {
        if(type == Constants.ListTypes.GENDER){
            genderSelection = position;
            if(genderSelection == 0)
                genderText.setTextColor(getResources().getColor(R.color.dustyWhite));
            else
                genderText.setTextColor(getResources().getColor(R.color.white));
            genderText.setText(genderArray[genderSelection]);
            genderText.setVisibility(View.VISIBLE);
            genderChoice.setVisibility(View.GONE);
        }
    }

}

File: app/src/main/java/com/example/policyfolio/ui/bottomsheets/InfoBottomSheet.kt
package com.example.policyfolio.ui.bottomsheets


import android.annotation.SuppressLint
import android.app.DatePickerDialog
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.view.isGone
import androidx.core.view.isVisible

import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.example.policyfolio.ui.adapters.ListAdapters.BasicDropdownTextAdapter
import com.example.policyfolio.util.Constants
import com.example.policyfolio.R
import com.example.policyfolio.viewmodels.HomeViewModel
import com.hbb20.CountryCodePicker
import com.wajahatkarim3.easyvalidation.core.view_ktx.validator

import java.util.Calendar

/**
 * A simple [Fragment] subclass.
 */
class InfoBottomSheet : Fragment, BasicDropdownTextAdapter.ParentCallback {

    private val callback: InfoSheetCallback
    private val viewModel: HomeViewModel
    private var rootView: View? = null

    private var name: EditText? = null
    private var email: EditText? = null
    private var phone: LinearLayout? = null
    private var phoneText: EditText? = null
    private var ccp: CountryCodePicker? = null
    private var birthday: TextView? = null
    private var genderText: TextView? = null
    private var genderChoice: RecyclerView? = null
    private var city: EditText? = null
    private var save: Button? = null

    private var nameError: TextView? = null
    private var emailError: TextView? = null
    private var phoneError: TextView? = null
    private var birthdayError: TextView? = null
    private var cityError: TextView? = null

    private var birthdayEpoch: Long? = null
    private var genderAdapter: BasicDropdownTextAdapter? = null
    private var genderArray: Array<String>? = null
    private var genderSelection: Int = 0

    @SuppressLint("ValidFragment")
    constructor(callback: InfoSheetCallback, viewModel: HomeViewModel) {
        this.callback = callback
        this.viewModel = viewModel
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.bottom_sheet_info, container, false)

        name = rootView!!.findViewById(R.id.name)
        email = rootView!!.findViewById(R.id.email)
        phoneText = rootView!!.findViewById(R.id.phone_text)
        phone = rootView!!.findViewById(R.id.phone)
        ccp = rootView!!.findViewById(R.id.ccp)
        birthday = rootView!!.findViewById(R.id.birthday)
        genderText = rootView!!.findViewById(R.id.gender_text)
        genderChoice = rootView!!.findViewById(R.id.gender_choice)
        city = rootView!!.findViewById(R.id.city)
        save = rootView!!.findViewById(R.id.save)

        nameError = rootView!!.findViewById(R.id.name_empty)
        emailError = rootView!!.findViewById(R.id.email_empty)
        phoneError = rootView!!.findViewById(R.id.phone_empty)
        birthdayError = rootView!!.findViewById(R.id.birthday_empty)
        cityError = rootView!!.findViewById(R.id.city_empty)

        ccp!!.registerCarrierNumberEditText(phoneText)
        ccp!!.setNumberAutoFormattingEnabled(true)

        setGenderAdapter()

        birthday!!.setOnClickListener { fetchDate() }

        save!!.setOnClickListener { save() }

        setAvailableInfo()

        return rootView
    }

    fun setAvailableInfo() {
        if (viewModel.email != null)
            email!!.setText(viewModel.email)
        if (viewModel.name != null)
            name!!.setText(viewModel.name)
        if (viewModel.phone != null){
            ccp!!.fullNumber = viewModel!!.phone
        }
        if (viewModel.birthday != null && viewModel.birthday > 0) {
            birthday!!.text = Constants.Time.DATE_FORMAT.format(viewModel.birthday!! * 1000)
            birthdayEpoch = viewModel.birthday
            birthday!!.setTextColor(resources.getColor(R.color.white))
        }
        setValue(viewModel.gender, Constants.ListTypes.GENDER)
        if (viewModel.city != null)
            city!!.setText(viewModel.city)
    }

    private fun save() {
        var isComplete = true

        val phone = ccp!!.formattedFullNumber
        if(ccp!!.isValidFullNumber){
            phoneError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
            viewModel!!.phone = phone
        } else{
            phoneError!!.text = "Invalid Phone Number!"
            phoneError!!.setTextColor(resources!!.getColor(R.color.red))
            isComplete = false
        }

        name!!.validator()
                .nonEmpty()
                .addErrorCallback { message ->
                    nameError!!.text = message
                    nameError!!.setTextColor(resources!!.getColor(R.color.red))
                    isComplete = false
                }
                .addSuccessCallback {
                    nameError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
                }.check()

        if (birthdayEpoch == null) {
            birthdayError!!.setTextColor(resources!!.getColor(R.color.red))
            isComplete = false
        } else {
            birthdayError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
        }

        city!!.validator()
                .nonEmpty()
                .addErrorCallback { message ->
                    cityError!!.text = message
                    cityError!!.setTextColor(resources!!.getColor(R.color.red))
                    isComplete = false
                }
                .addSuccessCallback {
                    cityError!!.setTextColor(resources!!.getColor(android.R.color.transparent))
                }.check()

        if (isComplete) {
            viewModel.complete = true
            viewModel.name = name!!.text.toString()
            viewModel.birthday = birthdayEpoch
            viewModel.gender = genderSelection
            viewModel.city = city!!.text.toString()
            viewModel.phone = phone
            callback.updateInfo()
        } else {
            Toast.makeText(context,"Incomplete Information!",Toast.LENGTH_SHORT).show()
        }
    }

    private fun fetchDate() {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = System.currentTimeMillis()

        val dateSetListener = DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
            calendar.set(year, monthOfYear, dayOfMonth)
            birthdayEpoch = calendar.timeInMillis / 1000
            birthday!!.text = Constants.Time.DATE_FORMAT.format(birthdayEpoch!! * 1000)
            birthday!!.setTextColor(resources.getColor(R.color.white))
        }

        DatePickerDialog(context,R.style.MyDatePickerDialogTheme, dateSetListener,
                calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
                calendar.get(Calendar.DAY_OF_MONTH)).show()
    }

    private fun setGenderAdapter() {
        genderSelection = 0
        genderArray = resources.getStringArray(R.array.gender_array)
        genderAdapter = BasicDropdownTextAdapter(context, genderArray, this, Constants.ListTypes.GENDER, resources.getColor(R.color.colorPrimaryDarkest))
        genderChoice!!.adapter = genderAdapter
        genderChoice!!.layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)

        genderText!!.setOnClickListener {
            genderText!!.visibility = View.GONE
            genderChoice!!.visibility = View.VISIBLE
        }
    }

    override fun setValue(position: Int, type: Int) {
        if (type == Constants.ListTypes.GENDER) {
            genderSelection = position
            if (genderSelection == 0)
                genderText!!.setTextColor(resources.getColor(R.color.dustyWhite))
            else
                genderText!!.setTextColor(resources.getColor(R.color.white))
            genderText!!.text = genderArray!![genderSelection]
            genderText!!.visibility = View.VISIBLE
            genderChoice!!.visibility = View.GONE
        }
    }

}
